### ISSUE TITLE
Request: consolidate collections in Library View

### ISSUE THREAD

**Issue by Amitheous** [advplyr/audiobookshelf#47]
Similar to how plex allows collections to be grouped together when viewing titles(like the Kung Fu Panda series in the image below), it would be nice to be able to do the same here.
![Plex Collections library](https://user-images.githubusercontent.com/44774645/134109635-dd096c54-6372-4ad1-b358-e9fcd6643324.PNG)
e.

**Comment by advplyr** [comment]
This is similar to #36 but it covers more use-cases. I didn't know Plex had this option, it is not easy to find, and it only applies when sorting by title.

Do you think there should be another field added called "collection"?
If we allow grouping by series and collection there would be conflicts when a book is in a collection and a series, so we could only allow one or the other at a time.

Would sorting be handled the same way as Plex? Since a group only has a title to sort by, the only time you would see the group is when you are sorting by title?

**Comment by Budlyte** [comment]
I'm answering from back on my post on #36 because I think it's more pertinent here.

Actually, after looking at Raymond E. Feist's stuff, I realize that Collections would be different than "Tags".  Feist's stuff, according to Goodreads, is 3 books here 3 books there, and so on but really there's an order to read them all together. So a Collection would be a custom list of books in a certain order, created by the user.

This would actually handled at the Collection level, a new thing, not the book level, and fix your mentioned issue of only allowing a book in a Series or Collection at one time.
Collection just equals a custom list of books, sorted in an order. This info wouldn't live in the book's edit page, allowing for a book to be included in more than one Collection.

**Comment by Budlyte** [comment]
Ooooohhhh, I apologize, I thought this was the same as #36 where items get automatically grouped based on being part of the same series.  I honestly forgot Plex has that custom Collections feature.

Yeah, Collections would be different than Tags, it would be a custom Reading List. I didn't realize these exist in books until I looked at Feist's stuff, I should've expected it though given how common it is in Comics.

Yeah, the library view does look nicer with them on their own tab, rather than lumped together.

**Comment by advplyr** [comment]
Eh, I was a bit off with what I said. In Plex you can create collections. Those collections live in another tab. However, they have an option:
![image](https://user-images.githubusercontent.com/67830747/134245889-966f589f-dcc0-48a1-ba07-ce4e33a43c36.png)

If you choose to "Hide items" in collections, it will group them in the main library like shown in the initial post screenshot. However, it will only group them if you are sorting by Title.

This may be one of those things that I'm just over-thinking, but if this is rolled out sloppily it is going to be difficult to unwind.

**Comment by Budlyte** [comment]
Yeah, I think if added then the result of Plex's Disabled selection is easiest to navigate. A separate tab for Collections.

Hell, this all being the Audiobooks library, this could be on the left side with options for all the other filters (Genres, Tags, Series, Authors). And then follow suit for the eBooks when added in? You did say you were wanting to rebuild this functionality.  :)

![image](https://user-images.githubusercontent.com/4451365/134249277-86e3ba59-c22e-4163-92fe-2e844317a16d.png)

**Comment by advplyr** [comment]
Damn, you have a huge collection. I'm surprised you haven't hit the limits on the http requests that send back an array of all your audiobooks.

I like where you are going with this. What do you think of using a "navigation rail" like this:

![image](https://user-images.githubusercontent.com/67830747/134349763-3f37abb3-a19a-47c8-85fc-42d3bc638f93.png)

If we do a side nav I want it to be as unobtrusive as possible

**Comment by advplyr** [comment]
I put something together real quick to try it out.

Would like to look for better icons if we go this route, or leave out the icons, I just think they are more eye-catching.

![image](https://user-images.githubusercontent.com/67830747/134357457-915b5fcc-1cee-41cf-9d8c-e3510b96baba.png)

**Comment by Budlyte** [comment]
Oh yeah, that looks really nice. I think more planning ahead may be needed though.

You mentioned in #50 that you would like to support eBooks as well, at some point, so taking a step further back means you'll need to support multiple libraries at some point. What does displaying multiple libraries look like down that road? 

Is it a layout like Plex or Komga, where the libraries are listed on the left with a "Home" or All view at the top?


![image](https://user-images.githubusercontent.com/4451365/134366044-77b7ec1e-cd07-494b-8377-97525d87ad5d.png)

![image](https://user-images.githubusercontent.com/4451365/134364517-232aa374-ea76-42de-b118-45d340122520.png)

![image](https://user-images.githubusercontent.com/4451365/134366197-f96c51c3-ef37-4394-9159-292d0e80f266.png)

![image](https://user-images.githubusercontent.com/4451365/134365372-556c2598-e958-433b-ab36-5590c6969fee.png)

Or is keeping it simple the best approach, simply have a toggle between the two libraries, and basically use the layout in your image above for both of them?

I'm afraid if you head down this road of changing layouts too early it may be more difficult to implement multiple libraries later.

**Comment by advplyr** [comment]
The ebook and audiobook can have the same details and cover art and live together in the same library.
We don't need to separate the 2 mediums like Plex does.

I couldn't get this looking nice, but it could be as simple as a dropdown.

![image](https://user-images.githubusercontent.com/67830747/134376921-69a78248-3dbd-4395-ac8a-e19840d19577.png)

Libraries can have different collections, tags, authors, etc.
Most importantly, libraries can have different user permissions and sharing.

It is good to start thinking of them now.

**Comment by Budlyte** [comment]
Oooh, yeah, library restrictions would allow things like Kids Audiobooks and such too.

I'm not asking that you build out for comics or anything, those are readily covered now. No one else is building an Audiobooks library server, so people are super grateful for your work in this.  I've seen your project mentioned on the Readarr discord lately.

I like your thoughts on combining the audio & print library, I guess is really the same way it was originally requested. The sidebar looks really nice.

**Comment by advplyr** [comment]
When viewing collections, series, etc. and you click on one of the items, is that just redirecting to the library with a filter applied?

Example: Click on the Enders Saga series -> viewing the library filtering for Enders Saga.

Something about that logic isn't making sense

**Comment by Budlyte** [comment]
Is that what it does in the background right now with the current dropdown menu? Part the goal here is to make the navigation to those more appealing, right? Move whatever those links are out to a larger screen view instead of that scrollbar.

**Comment by Amitheous** [comment]
Are you asking about how plex handles it? In that case clicking on the collection brings up a separate page with the collection contents. On mobile now so sorry if the format is strange but here's the same Kung Fu Panda example on mobile

![Screenshot_20210922-160254_Plex.jpg](https://user-images.githubusercontent.com/44774645/134428123-f07dd019-29ee-4ed1-b5ea-3d80bfe4f8f9.jpg)

![Screenshot_20210922-160304_Plex.jpg](https://user-images.githubusercontent.com/44774645/134428175-2f814f2c-b8ad-443b-b8ae-3512d82c2a7a.jpg)

**Comment by advplyr** [comment]
Yeah I like how Plex handles that. I'm not sure a separate page for collections & series would make sense with books.

With the side rail we were thinking of 2 different things. I wasn't thinking that the side rail would replace filtering, I was thinking that it would be a way to view groups.

On the collections page you would see a card for each collection with a cover, number of books, and name. Same with the other ones basically.
On the library page is where you see all the individual books, and you can filter and sort just like in a Plex library.

**Comment by Budlyte** [comment]
Oh, yeah I guess that is different. When you added everything into that side rail image I thought you were planning to move all those filter presets over to the left. Which seemed cool, because that scrolling list can be rough when you have a couple items in a category.

**Comment by advplyr** [comment]
I pushed an update this morning `v1.2.0` with the side rail. I didn't write release notes for that one.

I think we are on the right track but there is something missing, it doesn't feel right. I'm interested in your thoughts and critiques.
I want to get some more opinions on the side rail, it could be that it doesn't go well with the current design.

**Comment by Budlyte** [comment]
Hey, you're back!  I was starting to worry that we had broke you.

Are you just looking for critique on the sidebar right now?

I really like the ability to quickly switch between the different views with the sidebar, and what you did was stacking the images in the card looks good.

It is missing sorting, and I don't understand what it's doing by default. So once a Series is selected the books are in an undiscernible order.

Word wrap might be good for the rollover text. 
![image](https://user-images.githubusercontent.com/4451365/134747185-1a536b68-ebb7-44b1-ba67-7d92a7009b8b.png)

Could the Keyword Filter be added to this view?  That filter is suuuper useful, and I can see it being useful to filter down series by an Author if someone can't remember which specific series it was they're trying to find. Or the reverse, while in Author view using it to filter by a series name to see what other stuff they have.

A progress bar on the series card.

Basically if this view were still capable of all the same things the regular Library view is capable of, but with its nice graphical view. Maybe that's why it feels off?

Or maybe you're not liking how this corner is stitched together.   :P
![image](https://user-images.githubusercontent.com/4451365/134747906-c9711d98-4d3e-4c88-8b97-2d30805f3bf2.png)


And something is happening with the drawing, things have run off the screen and if the width is changed then the rows redraw to fewer and fewer until there is only one row running off of the screen.
![image](https://user-images.githubusercontent.com/4451365/134747223-e9da7836-0c77-4e70-b58f-e91795681121.png)

**Comment by Merijeek** [comment]
Nice to see I'm not the only fan of Skippy The Magnificent.

**Comment by advplyr** [comment]
> Hey, you're back! I was starting to worry that we had broke you.

No chance! I've been working on making this available outside of just Docker.

> Are you just looking for critique on the sidebar right now?

Critique everything, but yeah right now I'm conflicted on the side rail as far as the look and feel.

> I really like the ability to quickly switch between the different views with the sidebar, and what you did was stacking the images in the card looks good.

The stack of book covers for series I'm calling a "Group Card", as opposed to the regular book, "Book Card". The group card would be used anytime a group of books needs to be displayed on the bookshelf, so for series, collections, tags, etc.
I went with an easier version of the group card to start, I have a few other ideas for that.


> It is missing sorting, and I don't understand what it's doing by default. So once a Series is selected the books are in an undiscernible order.

Oops, forgot about that, I intended to use volume number order.

> Word wrap might be good for the rollover text.

Agree

> Could the Keyword Filter be added to this view? That filter is suuuper useful, and I can see it being useful to filter down series by an Author if someone can't remember which specific series it was they're trying to find. Or the reverse, while in Author view using it to filter by a series name to see what other stuff they have.

For sure. I only have a few series, but I can see how a collection of your size needs more control.

> A progress bar on the series card.
> 
> Basically if this view were still capable of all the same things the regular Library view is capable of, but with its nice graphical view. Maybe that's why it feels off?

I'm not sure what else other than a progress bar could go here. Maybe an option to batch edit the books in the series.

> Or maybe you're not liking how this corner is stitched together. :P

> And something is happening with the drawing, things have run off the screen and if the width is changed then the rows redraw to fewer and fewer until there is only one row running off of the screen.

That is a bad bug, I'll have to do this one first.

Thanks as always.

**Comment by advplyr** [comment]
Covered most of these updates in [v1.2.4](https://github.com/advplyr/audiobookshelf/releases/tag/v1.2.4)

- Keyword filter
- Progress bar
- Sort series by volume number
- Fix bookshelf overflow bug

![image](https://user-images.githubusercontent.com/67830747/134823357-ce9ec7de-2f18-4d91-9c48-60b4ade82f1c.png)

When clicking into a series, I added the volume number over the cover if it is set.

![image](https://user-images.githubusercontent.com/67830747/134823404-63d8855e-e3ad-4e7e-a248-2cd580125570.png)

**Comment by Budlyte** [comment]
`When clicking into a series, I added the volume number over the cover if it is set.
`
OMG, that's all kinds of sexy.

**Comment by Budlyte** [comment]
Hmmm,  well it doesn't freak out when I change the window size anymore but that's because it's always on one row that runs off the screen.
![image](https://user-images.githubusercontent.com/4451365/134824759-34a316a5-abc6-48b3-8c09-ff973c0a8873.png)

Also, not that it should reasonably happen, but I thought the progress bar was funny if books are marked read out of order. 

The keyword filter is missing too.

**Comment by advplyr** [comment]
Are you sure you updated? I definitely fixed the overflow. Check the version number on the settings page.

**Comment by Budlyte** [comment]
Will do.

Hitting "Select All" while in a Series selects the entire library.
![image](https://user-images.githubusercontent.com/4451365/134826928-a6f623fb-b48c-424a-86e2-bf1375651ae9.png)

**Comment by Budlyte** [comment]
Now it's somehow selecting all Series.

Showing number of Series
![image](https://user-images.githubusercontent.com/4451365/134912361-fd87aa70-04ec-440f-bc7c-0bdf9aff493e.png)

Book Selected
![image](https://user-images.githubusercontent.com/4451365/134912658-0763ab7d-3524-49ae-927e-5174d65b336d.png)

207 Series selected
![image](https://user-images.githubusercontent.com/4451365/134912699-4bfd7b1b-48c2-46e5-b682-1c949ea6b97b.png)

**Comment by advplyr** [comment]
I have this fixed but haven't published it yet. Looking for your valuable feedback on the new home page design in #63

**Comment by Amitheous** [comment]
Oh man I'm finally moved into the new place and got my server up and running again and I am amazed by the progress you made on this in such a short time. The implementation you came up with for the series is way better this way than how plex does it, at least for books. Fantastic work and keep it up!

**Comment by advplyr** [comment]
This has come a loooong way. Glad to close this out.
